### ISSUE TITLE
Drop 2.12 support

### ISSUE THREAD

**Issue by He-Pin** [apache/pekko#39]
Scala 2.12 support (e.g. ByteString has forked sources for 2.12 and 2.13 that lead to various bugs in the past), see also https://github.com/mdedetrich/akka-apache-project/discussions/27

_Originally posted by @jrudolph in https://github.com/mdedetrich/akka-apache-project/discussions/8#discussioncomment-3904101_
   
And now JDK 21 is released too.

**Comment by nvollmar** [comment]
I agree, it should be good enough to support the last two major versions

**Comment by He-Pin** [comment]
Yes, Scala 3.3 lts is near the corner too.

**Comment by spangaer** [comment]
Still on 2.12 ✅ 

Admittedly, we're not on the latest Akka either and we sure had the intention to move Scala 2.13 before doing so.

We're blocked from doing that before, but metals/VSCode is our route forward now.

Anyhow, am I wrong to state it's an Apache "thing" to support fairly "mature" runtimes?

Is it odd to suggest that dropping support should align with Scala center end-of-life of a version?

**Comment by SethTisue** [comment]
> Is it odd to suggest that dropping support should align with Scala center end-of-life of a version?

Scala 2.12.x itself would quite conceivably have been EOLed by now if it weren't for commercial sponsorship. So you can argue that for open source Pekko to drop 2.12 is actually consistent with the approach that Lightbend and the Center have taken with our stewardship of Scala 2.12.x — namely yes if commercially funded, otherwise perhaps no.

**Comment by spangaer** [comment]
I did do some research before writing that response, but it seems like these things aren't communicated very openly. (Till what you just wrote).
A published timeline for such things would really be helpful to base a decision on.
If not EOL, call it deep maintenance.

If it where Java, I'd think you'd target lts's and could start considering dropping 8.

**Comment by yishenggudou** [comment]
I feel that many people are still using version 2.12, including me

[Still on 2.12](https://trends.google.com/trends/explore?q=scala%202.12,scala%202.13&hl=en-US)

**Comment by nvollmar** [comment]
To get more facts, what is preventing you from upgrading to Scala 2.13?

**Comment by jrudolph** [comment]
I'd like to invite everyone to view issues like this from a community perspective. The whole first goal of this project is to preserve the existing status of the open source project and set up administrative structures to enable broader participation. So, there's no attempt to drop 2.12 from the initial release.

From there on, we'll see where it will go. In any case, it will require offers from the community to take on issues they see as pressing. Maintaining such a big project really is difficult and time consuming. I created these items originally, to show which current properties of the existing codebase make evolution more difficult (in this case, making sure everything works on at least 3 Scala versions can be a significant drain in productivity and often caused bugs).

IMO, the situation is pretty good even for 2.12 support: there will be an initial release supporting 2.12 and the basic organizational structures are set up that will allow the community to backport pressing issue to that first release branch and have some hope to get them released.

**Comment by mdedetrich** [comment]
I would disagree with dropping 2.12, even though technically speaking with Scala 3 its outside of the "latest 2 stable versions", adoption of Scala 3 is really slow and I wouldn't say its controversial that a lot of current Akka users would still be using and/or expecting 2.12 support.

In my opinion I would advocate for supporting 2.12 until it becomes painful to maintain it, then we can drop it.

**Comment by mdedetrich** [comment]
> It is painful for pekko-http as soon as we merge the Scala 3 branch.

I guess this would be because we would rely on Scala 3 to provide Scala 2.13 support so we would only have a single source directory?

**Comment by pjfanning** [comment]
Even if we decide to drop Scala 2.12 support for http module, we could keep it for other modules where the maintenance would be less of an issue.

**Comment by jrudolph** [comment]
> In my opinion I would advocate for supporting 2.12 until it becomes painful to maintain it, then we can drop it.

IMO, pain does not really work as a good metric because the pains will usually be small but consistent. E.g. just having that third version in the test and release matrix, it will add 50% more cost, possibility for flaky tests, missed test coverage, extra reasons for mima to fail etc. On the other hand, 2.13 has been released more than 3 years ago and has been super stable since a long while. There's little reason to stay on 2.12, so why should we?

**Comment by jrudolph** [comment]
> > It is painful for pekko-http as soon as we merge the Scala 3 branch.
> 
> I guess this would be because we would rely on Scala 3 to provide Scala 2.13 support so we would only have a single source directory?

Because, we would have to support hacky support structures to make it work at all and because of the general problems and costs of supporting an extra version in such a big project.

**Comment by mdedetrich** [comment]
> There's little reason to stay on 2.12, so why should we?

I can't answer this, I have no idea how much people are currently running akka-http on 2.12 which could cause a problem or whether we even care.

> Because, we would have to support hacky support structures to make it work at all and because of the general problems and costs of supporting an extra version in such a big project.

Yeah the reason I asked this is whether we have pekko-http already has separate source folders for `scala-2.12`/`scala-2.13`. I just had a look and `akka-http-core` appears to have such a structure. My original assumption is that with Scala 3 support, one would have to have a `scala-2.13` along with a `scala-3` folder and hence keeping the `scala-2.12` folder wouldn't be that much effort as long as too many things don't change in core of pekko-http (this is goes back to the "pain" part before).

If it was just a `scala-3` folder which would back compile to Scala 2.13 then that to me a much stronger argument since the mere presence of other source folders is to me the main complicating factor but I may be wrong here (and also my tolerance for pain might be higher).

The other part which might influence that just came to mind is the rest of the ecosystem as a whole. For example Kafka currently supports 2.12 and 2.13, I am not sure if they plan to support 2.12/2.13 and 3 when scala 3 is ready but since we have connectors (i.e. pekko-connectors-kafka) it might be problematic that Kafka itself is supporting older versions than pekko itself but I may be exaggerating things here.

**Comment by nvollmar** [comment]
Scala 2.13 has compiler flag to support certain Scala 3 syntax. Scala 3.0 supports both old and new styles, but newer Scala 3 versions will going to drop that. So mid to long term only Scala 2.13 in addition to Scala 3 will be viable to support without cloning the code base for Scala 2.12.

**Comment by nvollmar** [comment]
And on the other hand, migrating from Akka to Pekko will be more effort/risker than upgrading from Scala 2.12 to Scala 2.13 from my experience.

**Comment by spangaer** [comment]
Well, before we had standardized on ScalaIDE, which meant that we were stuck on Scala 2.12 since 2018 😔 

We've started adopting VSCode/Metals, but that setup still needs plenty of refinement, so only a couple of services use that already. It does look like this will be our way forward to Scala 2.13 and up.

But yes, it's unlikely that we bother to adopt Pekko before we can switch to 2.13, so I concur with many of the above statements.

To generalize, I actually think it's fair to consider dropping 2.12.

I just notice that Scala 2.12 isn't that dead yet. There's the Kafka example from above and afaik, writing an SBT plugin is also still 2.12.x.

**Comment by mdedetrich** [comment]
> I just notice that Scala 2.12 isn't that dead yet. There's the Kafka example from above and afaik, writing an SBT plugin is also still 2.12.x.

The main thing I am afraid of is jumping the gun when dropping 2.12 support even though a lot of other major Scala OS projects still support it (for various reasons), so at least in my opinion the optics of dropping 2.12 support in light of the rest of the Scala ecosystem wouldn't look the best. The only counter examples I can think of are projects like Flink, who have much bigger pains in supporting multiple Scala versions due to them using Maven (and in any case they are trying to completely phase out Scala as we speak so they aren't the best example to use).

Heck, even as you pointed out with SBT still on 2.12, unless SBT 2.x comes out then historically SBT kind of creates a lower floor for what is reasonable to support in terms of Scala open source libraries (at least if we are dealing with Scala libraries which are used for sbt plugins).

**Comment by pjfanning** [comment]
This is going to be contentious. I think we would need a community vote before proceeding.

**Comment by mdedetrich** [comment]
> This is going to be contentious. I think we would need a community vote before proceeding.

Agreed.

Also to add more to the debate, when pekko-http gets released I plan on adding pekko-http support to [scalaxb](https://github.com/eed3si9n/scalaxb) and since scalaxb is an sbt plugin (sbt plugins are built with Scala 2.12 only) it would essentially mean that scalaxb would be stuck with pekko-http 1.0.x series of updates.

While this may not be the end of the world, since sbt 1.x will be around for a while it would essentially mean that scalaxb's pekko-http backend wouldn't get new feature requests since that will only be for the 1.1.x series.

**Comment by pjfanning** [comment]
I think we should start by agreeing how long we will maintain the 1.0.x release lines. If we agree to make critical fixes to 1.0.x release lines, then I think we can make big changes in 1.1.x lines. People who can't use 1.1.x release lines because of the changes can continue to use 1.0.x releases.

**Comment by mdedetrich** [comment]
> I think we should start by agreeing how long we will maintain the 1.0.x release lines. 

I would rephrase the question by asking how much effort rather than how long. Since the agreement of 1.0.x is only backport of security/critical fixes we could "support" 1.0.x for a while assuming that these fixes are trivial to backport. If that stops becoming the case then we can consider dropping 1.0.x support.

The reasoning behind this is that our treatment from 1.0.x is special in the sense that its designed to be a bridge for people on Akka 2.6 and in this regard its not a "normal" release branch, i.e. the reason for its special status is the Akka fork. Hence our "normal" release branch is essentially 1.1.x, which means that major changes like dropping 2.12 should happen in 1.2.x.

**Comment by jtjeferreira** [comment]
> I plan on adding pekko-http support to scalaxb and since scalaxb is an sbt plugin (sbt plugins are built with Scala 2.12 only

Scalaxb SBT plugin is just generating code. It does not need to run pekko-http in scala 2.12...

**Comment by He-Pin** [comment]
https://spark.apache.org/releases/spark-release-4-0-0.html
Just dropped Scala 2.12

**Comment by raboof** [comment]
Kafka has also dropped 2.12 support in their 4.x major version it seems

**Comment by mdedetrich** [comment]
In general I am not against dropping Scala 2.12  but there are 2 points I would raise

1. Afaik (and I would like this to be confirmed), Scala 2.12 version of Pekko is still the most downloaded version of Pekko. The situation with this is even worse than with JDK 1.8, who appears to only have a few users using it (albeit those uses are quite critical i.e. enterprise). We should investigate how and why this is the case.
2. I wouldn't overblow the amount of effort required to support Scala 2.12. Pretty much every difference between the Scala versions has already been accounted for with our Pekko utils which abstracts over all Scala versions and the biggest divergence we have is between Scala 3 and Scala 2. Unlike JDK 1.8 with all of the various sbt build workarounds/Java 9 bespoke module stuff we have to deal with (which doesn't exist in JDK11+), Scala 2.12 only has some slightly different calls for specific library components. tl;dr, maintenance burden for Scala 2.12 is a weak argument all things considered.

**Comment by pjfanning** [comment]
repository.apache.org stats for June 2025
* pekko-actor_2.13 400k downloads
* pekko-actor_2.12 165k
* pekko-actor_3 104k

Aug 2024 (oldest stats still kept)
* pekko-actor_2.13 276k downloads
* pekko-actor_2.12 158k
* pekko-actor_3 55k

Tide moving against Scala 2.12.

It is also my view that Scala 2.12 users are typically maintaining old apps and usually not interested in upgrading libs except maybe to get a patch release.

**Comment by mdedetrich** [comment]
Thats good news then, I would suggest we track the Scala 2.12 usage while working on 2.0.x and if it continues to plummet hard proportionally to other Scala versions (which seems to be the case) then we can easily drop it.

**Comment by He-Pin** [comment]
I think we should drop 2.12.x and make Pekko works with Scala 2.13 3.3 and 3.9.

**Comment by He-Pin** [comment]
How about we do deprecations in 2.0.x but still keep Scala 2.12.x. Once we're done that, we release 2.0.x, and in 2.1.x , we totally drop 2.12.x support?

Then in the pekko 2.1.x , we cross compile pekko to Scala 2.13.x, 3.3.x and 3.9.x

**Comment by pjfanning** [comment]
With Scala 3, the artifactId gets a `_3` suffix. How can we support releasing jars for 3.3 and 3.9 when the artifactIds will be the same?

**Comment by He-Pin** [comment]
That may be changed one 3.9 is out

**Comment by SethTisue** [comment]
3.9 will still use the `_3` suffix. We don't anticipate that any library authors will crosspublish for 3.3 and 3.9.

The recommended approach will be:
* if your project works on 3.3 and you have a strong desire to support 3.3 users, publish on 3.3; it will also work on 3.9
* if your project actually requires something in 3.9, or if you simply want to move on and not be bothered with 3.3 anymore, publish on 3.9
* even if your main development branch moves to 3.9, you might also consider staying on 3.3 on a maintenance branch that only gets critical fixes

**Comment by He-Pin** [comment]
That's weird, then how can a user know if a library is published with 3.9 or 3.3?

**Comment by SethTisue** [comment]
> then how can a user know if a library is published with 3.9 or 3.3?

* from the library's release notes;
* or from the library's Scala 3 stdlib dependency in the POM (which I presume will also show in Scaladex);
* or at compile time, because the compiler will error if they try to use a 3.9 library on 3.3

the artifact suffix is used to indicate a break in backward compatibility (i.e., old libraries stop working); but 3.9 will be backward compatible with 3.3, and old libraries still work, so there's no need to change the artifact suffix

**Comment by mdedetrich** [comment]
I remember that there was some bugs with 3.3 that could not be fixed in the 3.3 series, but I think it had to do with `Unsafe` which we actually dropped in this series.

Aside from that, unless there are some compelling features in 3.9 it sounds like we should stick to 3.3

**Comment by He-Pin** [comment]
`Scala 3.3 will continue to receive support for at least a year after 3.9 is out, giving the community time to migrate the ecosystem.`

**Comment by He-Pin** [comment]
@pjfanning do you have any numbers about the 1.2.0 release? I really want to drop the Scala 2.12.x support in 2.0.0.

**Comment by pjfanning** [comment]
Maybe it's time to hold a vote on dev@pekko.apache.org.
I've emailed the dev list to notify mailing list subscribers that this is under discussion.
https://lists.apache.org/thread/gbojh6wcy94rzrno8xdt62ohx4qryk0f

**Comment by pjfanning** [comment]
I've opened a vote thread. You will need to reply to the email to register a vote.
https://lists.apache.org/thread/xol8o9tyl0pmdpffocwotv6x32hzmdw2

**Comment by SethTisue** [comment]
> I did do some research before writing that response, but it seems like these things aren't communicated very openly. (Till what you just wrote). A published timeline for such things would really be helpful to base a decision on. If not EOL, call it deep maintenance.

@spangaer I hope this is considered remedied by https://www.scala-lang.org/blog/2024/12/16/scala-2-maintenance.html and https://www.scala-lang.org/development/ — the phrasing we’re using for 2.12 is “minimal maintenance”

**Comment by SethTisue** [comment]
I don't feel I should actually vote since I'm not a Pekko (or Akka) user, but regardless, I want to firmly express my support for this change. If any Pekko users are still stuck on 2.12 for whatever reason, their project is a legacy project and they should be fine continuing to use a legacy Pekko version, _especially_ if the door remains open for that legacy version to get the occasional critical patch, e.g. for a security vulnerability. With that door open, then I feel like this change shouldn't even be considered controversial.

**Comment by He-Pin** [comment]
Yes, many issues can only be fixed in Scala 3. I think 2.13 has better type inference and support, and Scala 3 is superior. At work, we've been using 3.3.6 to compile our internal implementations. Scala 2.12 has caused significant problems for both CI and development and maintenance. If Akka were maintaining 2.12, I'd understand the commercial support, but for Pekko, as an open source project, I think the cost is too high. After all, dealing with 2.12 issues on weekends and nights is not a pleasant experience. For me, creating a better Java API experience probably has a higher ROI than supporting legacy 2.12 projects.

**Comment by pjfanning** [comment]
We get regular issues where PRs build ok except in Scala 2.12 and then need to be hacked to get the compilation sorted out. This adds to the workload for everyone. The end result might be less performant code for everyone as the change sometimes means additional checks or conversions.
We will save one third of our CI build workload with all the trees and water that will save.

**Comment by He-Pin** [comment]
Yes, especially since I've found that regular contributors don't actually use Scala 2.12 in their own work. For example, I use Pekko for my Java projects, but I always use Scala 3. If every compilation requires testing against 2.12, it's a waste of time. Life is short, so perhaps we could spend that time on a better Java experience and getting our projects approved by the Scala 3 community build.

eg：
```xml
        <dependency>
            <groupId>org.apache.pekko</groupId>
            <artifactId>pekko-stream_3</artifactId>
        </dependency>
```

**Comment by pjfanning** [comment]
this is done on main branch
